Mifare.c

```diff
@@ -1,203 +1,406 @@
+    CHECK_STATUS(status);
+
+
+    /* Initialize the discover component */
+    status = phacDiscLoop_Sw_Init(&sDiscLoop, sizeof(phacDiscLoop_Sw_DataParams_t), &sHal_Nfc_Ic.sHal);
+    CHECK_SUCCESS(status);
+
+    /* Load profile for Discovery loop */
+    LoadProfile();
+
+    status = phalMfc_Sw_Init(&salMfc, sizeof(phalMfc_Sw_DataParams_t), &spalMifare, NULL);
+    CHECK_STATUS(status);
+
+    /* Read the version of the reader IC */
+#if defined NXPBUILD__PHHAL_HW_RC523
+    status = phhalHw_Rc523_ReadRegister(&sHal_Nfc_Ic.sHal, PHHAL_HW_RC523_REG_VERSION, &bDataBuffer[0]);
+    CHECK_SUCCESS(status);
+#endif
+#if defined NXPBUILD__PHHAL_HW_RC663
+    status = phhalHw_Rc663_ReadRegister(&sHal_Nfc_Ic.sHal, PHHAL_HW_RC663_REG_VERSION, &bDataBuffer[0]);
+    CHECK_SUCCESS(status);
+#endif
 
-    /* Reset the RF field */
-    if ((status = phhalHw_FieldReset(&self->data.hal)) != PH_ERR_SUCCESS) {
-        return PyErr_Format(SelectError, "FieldReset command failed: %04x", status);
-    }
+    /* Return Success */
+    return PH_ERR_SUCCESS;
+}
 
-    /* Apply the type A protocol settings
-    * and activate the RF field. */
-    if ((status = phhalHw_ApplyProtocolSettings(&self->data.hal, PHHAL_HW_CARDTYPE_ISO14443A)) != PH_ERR_SUCCESS) {
-        return PyErr_Format(SelectError, "ApplyProtocolSettings command failed: %04x", status);
-    }
 
-    /* Empty the pAtqa */
-    memset(pAtqa, '\0', 2);
+PyObject *Mifare_init(Mifare *self, PyObject *args, PyObject *kwds) {
+	
+	int ret;
+	ret = Set_Interface_Link();
+	if(ret != 0)
+	{
+		return PyErr_Format(ReadError, "Hardware Initialisaton failed: %04x", ret);
+	}
+	Reset_reader_device();
+
+	ret = NfcRdLibInit();
+		if(ret != 0)
+	{
+		return PyErr_Format(ReadError, "Hardware Initialisaton failed: %04x", ret);
+	}
+    Py_RETURN_NONE;
+}
 
-    if ((status = phpalI14443p3a_RequestA(&self->data.I14443p3a, pAtqa)) != PH_ERR_SUCCESS) {
-        return PyErr_Format(SelectError, "RequestA command failed: %04x", status);
-    }
 
-    /* Reset the RF field */
-    if ((status = phhalHw_FieldReset(&self->data.hal)) != PH_ERR_SUCCESS) {
-        return PyErr_Format(SelectError, "FieldReset command failed: %04x", status);
-    }
-    /* Activate the communication layer part 3 of the ISO 14443A standard. */
-    if ( phpalI14443p3a_ActivateCard(&self->data.I14443p3a, NULL, 0x00, byteBuffer, &byteBufferSize, &bSak, &bMoreCardsAvailable) == PH_ERR_SUCCESS) {
-        // Card is present, return it as a python object.
-        uint8_t i;
-        char asciiBuffer[UID_ASCII_BUFFER_SIZE];
-
-        if (byteBufferSize + 1 > UID_ASCII_BUFFER_SIZE) {
-            // Truncate if we got back too much data
-            byteBufferSize = UID_ASCII_BUFFER_SIZE - 1;
-        }
+PyObject *Mifare_select(Mifare *self)
+{
+phStatus_t  status = 0;
+    uint16_t    wTagsDetected = 0;
+    uint8_t     bUid[PHAC_DISCLOOP_I3P3A_MAX_UID_LENGTH];
+    uint8_t     bUidSize;
+    while(1)    /* Continuous loop */
```

Comment:
`select()` should NOT be blocking.  If an error occurs or no tag is present, an error should be returned like this: `return PyErr_Format(SelectError, "Error message here");`